<!DOCTYPE html>

<html lang="python" data-content_root="../">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>Blast Documentation &#8212; OrthoEvolution 1.0.0b2 documentation</title>
    <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=fa44fd50" />
    <link rel="stylesheet" type="text/css" href="../_static/basic.css?v=b08954a9" />
    <link rel="stylesheet" type="text/css" href="../_static/alabaster.css?v=27fed22d" />
    <script src="../_static/documentation_options.js?v=3e4b6f43"></script>
    <script src="../_static/doctools.js?v=9bcbadda"></script>
    <script src="../_static/sphinx_highlight.js?v=dc90522c"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
   
  <link rel="stylesheet" href="../_static/custom.css" type="text/css" />
  

  
  

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <section id="blast-documentation">
<h1>Blast Documentation<a class="headerlink" href="#blast-documentation" title="Link to this heading">¶</a></h1>
<p>This module uses <a class="reference external" href="https://blast.ncbi.nlm.nih.gov/Blast.cgi?PAGE_TYPE=BlastDocs&amp;DOC_TYPE=Download">NCBI’s standalone
blast</a>
to generate blastn results. The results are parsed for the best hit,
which are used to get accession numbers.</p>
<section id="what-is-blast">
<h2>What is BLAST?<a class="headerlink" href="#what-is-blast" title="Link to this heading">¶</a></h2>
<p>Per NCBI, the <a class="reference external" href="https://blast.ncbi.nlm.nih.gov/Blast.cgi">Basic Local Alignment Search Tool
(BLAST)</a> finds regions of
local similarity between sequences. The program compares nucleotide or
protein sequences to sequence databases and calculates the statistical
significance of matches. BLAST can be used to infer functional and
evolutionary relationships between sequences as well as help identify
members of gene families.</p>
<p>We use NCBI’s <code class="docutils literal notranslate"><span class="pre">blastn</span></code> task to generate a best hit in order to infer
orthology which is under the umbrella of comparative genetics/genomics
Comparative genetics/genomics is a field of biological research in which
the genome sequences of different species human, mouse, and a wide
variety of other organisms from bacteria to chimpanzees are compared.</p>
<p>Using this package, we compared these
<a class="reference external" href="http://www.guidetopharmacology.org/targets.jsp">genes</a> of interest
across a group of
<a class="reference external" href="ftp://ftp.ncbi.nlm.nih.gov/genomes/refseq/vertebrate_mammalian/">species</a>.</p>
<section id="how-do-we-configure-and-run-blast">
<h3>How do we configure and run blast?<a class="headerlink" href="#how-do-we-configure-and-run-blast" title="Link to this heading">¶</a></h3>
<p>Running blast is the most complex aspect of this package, but we’ve
found a way to simplify the <strong>automation of blasting</strong> while also
<strong>limiting blast searches by taxonomy id</strong>.</p>
<p>Before you use this function, you need for <code class="docutils literal notranslate"><span class="pre">NCBI</span> <span class="pre">Blast+</span></code> to be
installed and in your path. Download the latest standalone blast
executables from
<a class="reference external" href="ftp://ftp.ncbi.nlm.nih.gov/blast/executables/blast+/LATEST/">here</a>.
We are currently using version <code class="docutils literal notranslate"><span class="pre">2.8.1</span></code>.</p>
<section id="our-blast-methods">
<h4>Our Blast Methods<a class="headerlink" href="#our-blast-methods" title="Link to this heading">¶</a></h4>
<p>NCBI’s <code class="docutils literal notranslate"><span class="pre">blastn</span></code> can be configured (using its parameters) in a number
of different ways (i.e. local or remote use and with seqidlists or
taxids). For typical orthology analyses, it’s important to take
advantage of the speed and efficiency of NCBI’s newest preformatted
blast databases
(<a class="reference external" href="https://ftp.ncbi.nlm.nih.gov/blast/db/v5/">blastdbv5</a>). In order to
do that, we’ve implemented a method (<code class="docutils literal notranslate"><span class="pre">1</span></code>) that uses taxids (taxonomic
groups — species level and higher level taxa). View more about our
methods below.</p>
<table class="docutils align-default">
<thead>
<tr class="row-odd"><th class="head"><p>Method</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>1</p></td>
<td><p>Local blast using taxids. Utilizes local databases
(<code class="docutils literal notranslate"><span class="pre">refseq_rna_v5</span></code>).</p></td>
</tr>
<tr class="row-odd"><td><p>2</p></td>
<td><p>Remote blast using an entrez query. Uses entrez species
name and query</p></td>
</tr>
<tr class="row-even"><td><p>None</p></td>
<td><p>A single query method not useful for orthology inference</p></td>
</tr>
</tbody>
</table>
</section>
<section id="our-custom-accession-file-format">
<h4>Our Custom Accession File Format<a class="headerlink" href="#our-custom-accession-file-format" title="Link to this heading">¶</a></h4>
<p>We use a specifically formatted <code class="docutils literal notranslate"><span class="pre">accession</span> <span class="pre">file</span></code> with our headers as
<code class="docutils literal notranslate"><span class="pre">Tier</span></code>, <code class="docutils literal notranslate"><span class="pre">Gene</span></code>, <code class="docutils literal notranslate"><span class="pre">Organism</span></code> to store blast output and input. This
allows for distinguishing genes by families or features. The <code class="docutils literal notranslate"><span class="pre">Tier</span></code>
header can be omitted, but the other headers are requirements. The
Accession numbers are stored in a <code class="docutils literal notranslate"><span class="pre">.csv</span></code> file. The following table is
an example of how we format our blast input file.</p>
<table class="docutils align-default">
<thead>
<tr class="row-odd"><th class="head"><p>Tier</p></th>
<th class="head"><p>Gene</p></th>
<th class="head"><p>Homo_sapiens</p></th>
<th class="head"><p>Macaca_mulatta</p></th>
<th class="head"><p>Mus_musculus</p></th>
<th class="head"><p>Rattus_norvegicus</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>1</p></td>
<td><p>ADRA1A</p></td>
<td><p>NM_000680.3</p></td>
<td></td>
<td></td>
<td></td>
</tr>
<tr class="row-odd"><td><p>2</p></td>
<td><p>ADRA1B</p></td>
<td><p>NM_000679.3</p></td>
<td></td>
<td></td>
<td></td>
</tr>
<tr class="row-even"><td><p>3</p></td>
<td><p>ADRA1D</p></td>
<td><p>NM_000678.3</p></td>
<td></td>
<td></td>
<td></td>
</tr>
<tr class="row-odd"><td><p>4</p></td>
<td><p>ADRA2A</p></td>
<td><p>NM_000681.3</p></td>
<td></td>
<td></td>
<td></td>
</tr>
<tr class="row-even"><td><p>Immune</p></td>
<td><p>ADRA2B</p></td>
<td><p>NM_000682.6</p></td>
<td></td>
<td></td>
<td></td>
</tr>
<tr class="row-odd"><td><p>Addiction</p></td>
<td><p>CHRM1</p></td>
<td><p>NM_000738.2</p></td>
<td></td>
<td></td>
<td></td>
</tr>
<tr class="row-even"><td><p>Ugly</p></td>
<td><p>CHRM2</p></td>
<td><p>NM_000739.2</p></td>
<td></td>
<td></td>
<td></td>
</tr>
<tr class="row-odd"><td><p>Other</p></td>
<td><p>CHRM3</p></td>
<td><p>NM_000740.2</p></td>
<td></td>
<td></td>
<td></td>
</tr>
<tr class="row-even"><td><p>GPCR</p></td>
<td><p>CHRM5</p></td>
<td><p>NM_012125.3</p></td>
<td></td>
<td></td>
<td></td>
</tr>
<tr class="row-odd"><td><p>Isoforms</p></td>
<td><p>CNR1</p></td>
<td><p>NM_016083.4</p></td>
<td></td>
<td></td>
<td></td>
</tr>
</tbody>
</table>
<p>The .csv file requires some manual configuration, and, while tedious, it
is also currently fundamental for the API.</p>
<p>Below we have defined the headers:</p>
<ul class="simple">
<li><p><strong>Tier</strong>: The target genes need a ranking or categorization based on
the experiment. These can be user defined or a preset tier system can
be used. In the future the different tiers will allow the user to
control the order that each gene is processed.</p></li>
<li><p><strong>Gene</strong>: The genes are HGNC aliases for the target genes of interest.
In the future we will be able to process the HGNC .csv file to further
automate the creation of this template file.</p></li>
<li><p><strong>Query</strong>: The query organism is placed into the 3rd column of the
.csv file. In the example Homo sapiens is used. Each taxa is a string
in the format of “<em>Genus_species</em>”. The query organism also has to
have accession numbers for each gene. It is therefore highly important
to pick a well annotated species for accurate analysis.</p></li>
</ul>
</section>
</section>
</section>
<section id="examples">
<h2>Examples<a class="headerlink" href="#examples" title="Link to this heading">¶</a></h2>
<p>The main class to use is <code class="docutils literal notranslate"><span class="pre">OrthoBlastN</span></code> in order to run blast. In order
to run <code class="docutils literal notranslate"><span class="pre">OrthoBlastN</span></code> without using our database management features,
the <code class="docutils literal notranslate"><span class="pre">BLASTDB</span></code> paths must be set in your environment.</p>
<section id="performing-blast-post-blast-analysis">
<h3>Performing Blast &amp; Post-Blast Analysis<a class="headerlink" href="#performing-blast-post-blast-analysis" title="Link to this heading">¶</a></h3>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">OrthoEvol.Orthologs.Blast</span> <span class="kn">import</span> <span class="n">OrthoBlastN</span>


<span class="c1"># Use an existing list of gpcr genes</span>
<span class="n">gpcr_blastn</span> <span class="o">=</span> <span class="n">OrthoBlastN</span><span class="p">(</span><span class="n">project</span><span class="o">=</span><span class="s2">&quot;orthology-gpcr&quot;</span><span class="p">,</span> <span class="n">method</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>
                             <span class="n">save_data</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">acc_file</span><span class="o">=</span><span class="s2">&quot;gpcr.csv&quot;</span><span class="p">,</span>
                             <span class="n">copy_from_package</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

<span class="c1"># View the list of genes</span>
<span class="n">gpcr_blastn</span><span class="o">.</span><span class="n">gene_list</span>

<span class="c1"># View the blast dataframe</span>
<span class="n">gpcr_blastn</span><span class="o">.</span><span class="n">df</span>

<span class="c1"># Start the blast</span>
<span class="n">gpcr_blastn</span><span class="o">.</span><span class="n">run</span><span class="p">()</span>

<span class="c1"># Use your own accessions file.</span>
<span class="c1"># You don&#39;t need to copy from package to use your own genes</span>
<span class="n">my_blastn</span> <span class="o">=</span> <span class="n">OrthoBlastN</span><span class="p">(</span><span class="n">project</span><span class="o">=</span><span class="s2">&quot;orthology-project&quot;</span><span class="p">,</span> <span class="n">method</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>
                             <span class="n">save_data</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">acc_file</span><span class="o">=</span><span class="s2">&quot;mygenes.csv&quot;</span><span class="p">,</span>
                             <span class="n">copy_from_package</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

<span class="n">my_blastn</span><span class="o">.</span><span class="n">run</span><span class="p">()</span>
</pre></div>
</div>
</section>
<section id="customing-with-baseblastn">
<h3>Customing with BaseBlastN<a class="headerlink" href="#customing-with-baseblastn" title="Link to this heading">¶</a></h3>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">OrthoEvol.Orthologs.Blast</span> <span class="kn">import</span> <span class="n">BaseBlastN</span>

<span class="c1"># This is more pythonic with YAML loading</span>
<span class="n">blastconfig</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s2">&quot;project&quot;</span><span class="p">:</span> <span class="s2">&quot;test&quot;</span><span class="p">,</span>
    <span class="s2">&quot;method&quot;</span><span class="p">:</span> <span class="mi">1</span><span class="p">,</span>
    <span class="s2">&quot;taxon_file&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
    <span class="s2">&quot;go_list&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
    <span class="s2">&quot;post_blast&quot;</span><span class="p">:</span> <span class="kc">True</span><span class="p">,</span>
    <span class="s2">&quot;template&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
    <span class="s2">&quot;save_data&quot;</span><span class="p">:</span> <span class="kc">True</span><span class="p">,</span>
    <span class="s2">&quot;copy_from_package&quot;</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span>
    <span class="s2">&quot;acc_file&quot;</span><span class="p">:</span> <span class="s2">&quot;test_blast.csv&quot;</span><span class="p">,</span>
    <span class="s2">&quot;project_path&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
    <span class="s2">&quot;proj_mana&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
    <span class="s2">&quot;ref_species&quot;</span><span class="p">:</span> <span class="s2">&quot;Homo_sapiens&quot;</span>
<span class="p">}</span>


<span class="n">test_blast</span> <span class="o">=</span> <span class="n">BaseBlastN</span><span class="p">(</span><span class="o">**</span><span class="n">blastconfig</span><span class="p">)</span>
<span class="n">test_blast</span><span class="o">.</span><span class="n">configure</span><span class="p">(</span><span class="n">test_blast</span><span class="o">.</span><span class="n">blast_human</span><span class="p">,</span> <span class="n">auto_start</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</section>
</section>
</section>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="Main">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="../index.html">OrthoEvolution</a></h1>



<p class="blurb">An easy to use and comprehensive python package which aids in the analysis and visualization of comparative evolutionary genetics</p>




<p>
<iframe src="https://ghbtns.com/github-btn.html?user=Datasnakes&repo=OrthoEvolution&type=watch&count=true&size=large&v=2"
  allowtransparency="true" frameborder="0" scrolling="0" width="200px" height="35px"></iframe>
</p>






  <div>
    <h3><a href="../index.html">Table of Contents</a></h3>
    <ul>
<li><a class="reference internal" href="#">Blast Documentation</a><ul>
<li><a class="reference internal" href="#what-is-blast">What is BLAST?</a><ul>
<li><a class="reference internal" href="#how-do-we-configure-and-run-blast">How do we configure and run blast?</a><ul>
<li><a class="reference internal" href="#our-blast-methods">Our Blast Methods</a></li>
<li><a class="reference internal" href="#our-custom-accession-file-format">Our Custom Accession File Format</a></li>
</ul>
</li>
</ul>
</li>
<li><a class="reference internal" href="#examples">Examples</a><ul>
<li><a class="reference internal" href="#performing-blast-post-blast-analysis">Performing Blast &amp; Post-Blast Analysis</a></li>
<li><a class="reference internal" href="#customing-with-baseblastn">Customing with BaseBlastN</a></li>
</ul>
</li>
</ul>
</li>
</ul>

  </div>
<search id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" autocomplete="off" autocorrect="off" autocapitalize="off" spellcheck="false"/>
      <input type="submit" value="Go" />
    </form>
    </div>
</search>
<script>document.getElementById('searchbox').style.display = "block"</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &#169;2017, Shaurita Hutchins & Robert Gilmore.
      
      |
      Powered by <a href="https://www.sphinx-doc.org/">Sphinx 8.2.3</a>
      &amp; <a href="https://alabaster.readthedocs.io">Alabaster 1.0.0</a>
      
      |
      <a href="../_sources/orthologs/blastreadme.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    
    <a href="https://github.com/Datasnakes/OrthoEvolution" class="github">
        <img src="../_static/github-banner.svg" alt="Fork me on GitHub" class="github"/>
    </a>
    

    
  </body>
</html>