<!DOCTYPE html>

<html lang="python" data-content_root="../">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>Genbank Documentation &#8212; OrthoEvolution 1.0.0b2 documentation</title>
    <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=fa44fd50" />
    <link rel="stylesheet" type="text/css" href="../_static/basic.css?v=b08954a9" />
    <link rel="stylesheet" type="text/css" href="../_static/alabaster.css?v=27fed22d" />
    <script src="../_static/documentation_options.js?v=3e4b6f43"></script>
    <script src="../_static/doctools.js?v=9bcbadda"></script>
    <script src="../_static/sphinx_highlight.js?v=dc90522c"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
   
  <link rel="stylesheet" href="../_static/custom.css" type="text/css" />
  

  
  

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <section id="genbank-documentation">
<h1>Genbank Documentation<a class="headerlink" href="#genbank-documentation" title="Link to this heading">¶</a></h1>
<p>Retrieve genbank files and extract specific features sucha as <code class="docutils literal notranslate"><span class="pre">cds</span></code> or
<code class="docutils literal notranslate"><span class="pre">aa</span></code>. Also, write the features to text files.</p>
<p>If you haven’t worked with genbank files before or are unfamiliar with
what they look like, view a <a class="reference external" href="https://www.ncbi.nlm.nih.gov/Sitemap/samplerecord.html">sample genbank
record</a>.</p>
<section id="usage">
<h2>Usage<a class="headerlink" href="#usage" title="Link to this heading">¶</a></h2>
<p>The main class is <code class="docutils literal notranslate"><span class="pre">GenBank</span></code>.</p>
<section id="notable-file-formats">
<h3>Notable File Formats<a class="headerlink" href="#notable-file-formats" title="Link to this heading">¶</a></h3>
<table class="docutils align-default">
<thead>
<tr class="row-odd"><th class="head"><p>Extension</p></th>
<th class="head"><p>Meaning</p></th>
<th class="head"><p>Notes</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>fasta</p></td>
<td><p>generic
fasta</p></td>
<td><p>Any generic fasta file. Other extensions can be
fas, fa, seq, fsa</p></td>
</tr>
<tr class="row-odd"><td><p>fna</p></td>
<td><p>fasta
nucleic
acid</p></td>
<td><p>Used generically to specify nucleic acids.</p></td>
</tr>
<tr class="row-even"><td><p>ffn</p></td>
<td><p>FASTA
nucleotide
of gene
regions</p></td>
<td><p>Contains coding regions for a genome.</p></td>
</tr>
<tr class="row-odd"><td><p>faa</p></td>
<td><p>fasta
amino acid</p></td>
<td><p>Contains amino acids. A multiple protein fasta
file can have the more specific extension mpfa.</p></td>
</tr>
<tr class="row-even"><td><p>frn</p></td>
<td><p>FASTA
non-coding
RNA</p></td>
<td><p>Contains non-coding RNA regions for a genome,
in DNA alphabet e.g. tRNA, rRNA</p></td>
</tr>
</tbody>
</table>
</section>
</section>
<section id="examples">
<h2>Examples<a class="headerlink" href="#examples" title="Link to this heading">¶</a></h2>
<section id="perform-genbank-feature-extraction">
<h3>Perform Genbank Feature Extraction<a class="headerlink" href="#perform-genbank-feature-extraction" title="Link to this heading">¶</a></h3>
</section>
</section>
</section>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="Main">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="../index.html">OrthoEvolution</a></h1>



<p class="blurb">An easy to use and comprehensive python package which aids in the analysis and visualization of comparative evolutionary genetics</p>




<p>
<iframe src="https://ghbtns.com/github-btn.html?user=Datasnakes&repo=OrthoEvolution&type=watch&count=true&size=large&v=2"
  allowtransparency="true" frameborder="0" scrolling="0" width="200px" height="35px"></iframe>
</p>






  <div>
    <h3><a href="../index.html">Table of Contents</a></h3>
    <ul>
<li><a class="reference internal" href="#">Genbank Documentation</a><ul>
<li><a class="reference internal" href="#usage">Usage</a><ul>
<li><a class="reference internal" href="#notable-file-formats">Notable File Formats</a></li>
</ul>
</li>
<li><a class="reference internal" href="#examples">Examples</a><ul>
<li><a class="reference internal" href="#perform-genbank-feature-extraction">Perform Genbank Feature Extraction</a></li>
</ul>
</li>
</ul>
</li>
</ul>

  </div>
<search id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" autocomplete="off" autocorrect="off" autocapitalize="off" spellcheck="false"/>
      <input type="submit" value="Go" />
    </form>
    </div>
</search>
<script>document.getElementById('searchbox').style.display = "block"</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &#169;2017, Shaurita Hutchins & Robert Gilmore.
      
      |
      Powered by <a href="https://www.sphinx-doc.org/">Sphinx 8.2.3</a>
      &amp; <a href="https://alabaster.readthedocs.io">Alabaster 1.0.0</a>
      
      |
      <a href="../_sources/orthologs/genbankreadme.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    
    <a href="https://github.com/Datasnakes/OrthoEvolution" class="github">
        <img src="../_static/github-banner.svg" alt="Fork me on GitHub" class="github"/>
    </a>
    

    
  </body>
</html>