<!DOCTYPE html>

<html lang="python" data-content_root="../">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>Phylogenetics Documentation &#8212; OrthoEvolution 1.0.0b2 documentation</title>
    <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=fa44fd50" />
    <link rel="stylesheet" type="text/css" href="../_static/basic.css?v=b08954a9" />
    <link rel="stylesheet" type="text/css" href="../_static/alabaster.css?v=27fed22d" />
    <script src="../_static/documentation_options.js?v=3e4b6f43"></script>
    <script src="../_static/doctools.js?v=9bcbadda"></script>
    <script src="../_static/sphinx_highlight.js?v=dc90522c"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
   
  <link rel="stylesheet" href="../_static/custom.css" type="text/css" />
  

  
  

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <section id="phylogenetics-documentation">
<h1>Phylogenetics Documentation<a class="headerlink" href="#phylogenetics-documentation" title="Link to this heading">¶</a></h1>
<p>This documentation will provide information and guidelines about how we
use the Phylogenetics modules related to this package.</p>
<section id="overview">
<h2>Overview<a class="headerlink" href="#overview" title="Link to this heading">¶</a></h2>
<p>Phylogenetics is best defined as the study of evolutionary relationships
among biological entities. In our case, those entities are species. We
are seeking to learn how mammals (more specifically primates) compare to
each other given a group of genes (GPCRs and addiction related).</p>
<p>PAML in particular is the most rigorous software for helping us to
understand the potentially significant differences in genes across
different mammalian species. From there, we can decide which genes we
will further study in cell culture projects or assays.</p>
</section>
<section id="dependencies">
<h2>Dependencies<a class="headerlink" href="#dependencies" title="Link to this heading">¶</a></h2>
<p>It is critical to have these phylogenetic software installed and
available on your path in order to use the Phylogenetics submodules or
you can take a look at our <a class="reference external" href="https://github.com/datasnakes/external-apps">external apps
repository</a> to help you
install these software on your machine.</p>
</section>
<section id="modules-attributes-available">
<h2>Modules/Attributes available<a class="headerlink" href="#modules-attributes-available" title="Link to this heading">¶</a></h2>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">OrthoEvol.Orthologs</span> <span class="kn">import</span> <span class="n">Phylogenetics</span>

<span class="c1"># Find out what subclasses are available for use</span>
<span class="nb">dir</span><span class="p">(</span><span class="n">Phylogenetics</span><span class="p">)</span>

<span class="n">Out</span><span class="p">[</span><span class="mi">1</span><span class="p">]:</span>
<span class="p">[</span><span class="s1">&#39;AlignIO&#39;</span><span class="p">,</span>
 <span class="s1">&#39;ETE3PAML&#39;</span><span class="p">,</span>
 <span class="s1">&#39;FilteredTree&#39;</span><span class="p">,</span>
 <span class="s1">&#39;IQTree&#39;</span><span class="p">,</span>
 <span class="s1">&#39;IQTreeCommandline&#39;</span><span class="p">,</span>
 <span class="s1">&#39;OrthologsWarning&#39;</span><span class="p">,</span>
 <span class="s1">&#39;PhyML&#39;</span><span class="p">,</span>
 <span class="s1">&#39;Phylip&#39;</span><span class="p">,</span>
 <span class="s1">&#39;RelaxPhylip&#39;</span><span class="p">,</span>
 <span class="s1">&#39;TreeViz&#39;</span><span class="p">,</span>
 <span class="s1">&#39;__all__&#39;</span><span class="p">,</span>
 <span class="s1">&#39;__builtins__&#39;</span><span class="p">,</span>
 <span class="s1">&#39;__cached__&#39;</span><span class="p">,</span>
 <span class="s1">&#39;__doc__&#39;</span><span class="p">,</span>
 <span class="s1">&#39;__file__&#39;</span><span class="p">,</span>
 <span class="s1">&#39;__loader__&#39;</span><span class="p">,</span>
 <span class="s1">&#39;__name__&#39;</span><span class="p">,</span>
 <span class="s1">&#39;__package__&#39;</span><span class="p">,</span>
 <span class="s1">&#39;__path__&#39;</span><span class="p">,</span>
 <span class="s1">&#39;__spec__&#39;</span><span class="p">,</span>
 <span class="s1">&#39;warnings&#39;</span><span class="p">]</span>
</pre></div>
</div>
</section>
<section id="examples">
<h2>Examples<a class="headerlink" href="#examples" title="Link to this heading">¶</a></h2>
<p>In the beginning stages of our project, we tested various phylogenetic
programs to see which worked well for us.</p>
<p>In this module, we include classes and ways to use PAML, Phylip, PhyML,
IQTREE, and Biopython’s Bio.Phylo class.</p>
<section id="using-phyml-with-relaxphylip">
<h3>Using PhyML with RelaxPhylip<a class="headerlink" href="#using-phyml-with-relaxphylip" title="Link to this heading">¶</a></h3>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># Now you can import a class you want to utilize</span>
<span class="kn">from</span> <span class="nn">OrthoEvol.Orthologs.Phylogenetics</span> <span class="kn">import</span> <span class="n">PhyML</span><span class="p">,</span> <span class="n">RelaxPhylip</span>

<span class="n">RelaxPhylip</span><span class="p">(</span><span class="s2">&quot;HTR1A_aligned.fasta&quot;</span><span class="p">,</span> <span class="s2">&quot;HTR1A_aligned.phy&quot;</span><span class="p">)</span>

<span class="c1"># Generate a maximum likelihood tree from the phylip formatted alignment file.</span>
<span class="n">PhyML</span><span class="p">(</span><span class="s2">&quot;HTR1A_aligned.phy&quot;</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="using-ete3paml">
<h3>Using ETE3PAML<a class="headerlink" href="#using-ete3paml" title="Link to this heading">¶</a></h3>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">OrthoEvol.Orthologs.Phylogenetics.PAML</span> <span class="kn">import</span> <span class="n">ETE3PAML</span>

<span class="n">paml</span> <span class="o">=</span> <span class="n">ETE3PAML</span><span class="p">(</span><span class="n">alignmentfile</span><span class="o">=</span><span class="s1">&#39;.ffn&#39;</span><span class="p">,</span> <span class="n">speciestree</span><span class="o">=</span><span class="s1">&#39;.nw&#39;</span><span class="p">,</span> <span class="n">workdir</span><span class="o">=</span><span class="s1">&#39;&#39;</span><span class="p">)</span>

<span class="n">paml</span><span class="o">.</span><span class="n">run</span><span class="p">(</span><span class="n">pamlsrc</span><span class="o">=</span><span class="s1">&#39;path/to/codeml/binary&#39;</span><span class="p">,</span> <span class="n">output_folder</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="using-the-filteredtree-implementation-of-iqtree">
<h3>Using the FilteredTree implementation of IQTree<a class="headerlink" href="#using-the-filteredtree-implementation-of-iqtree" title="Link to this heading">¶</a></h3>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">OrthoEvol.Orthologs.Phylogenetics</span> <span class="kn">import</span> <span class="n">FilteredTree</span>

<span class="n">FilteredTree</span><span class="p">(</span><span class="n">alignment</span><span class="p">,</span> <span class="n">dataType</span><span class="o">=</span><span class="s1">&#39;CODON&#39;</span><span class="p">,</span> <span class="n">working_dir</span><span class="o">=</span><span class="s1">&#39;path/of/working/directory&#39;</span><span class="p">)</span>
</pre></div>
</div>
</section>
</section>
</section>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="Main">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="../index.html">OrthoEvolution</a></h1>



<p class="blurb">An easy to use and comprehensive python package which aids in the analysis and visualization of comparative evolutionary genetics</p>




<p>
<iframe src="https://ghbtns.com/github-btn.html?user=Datasnakes&repo=OrthoEvolution&type=watch&count=true&size=large&v=2"
  allowtransparency="true" frameborder="0" scrolling="0" width="200px" height="35px"></iframe>
</p>






  <div>
    <h3><a href="../index.html">Table of Contents</a></h3>
    <ul>
<li><a class="reference internal" href="#">Phylogenetics Documentation</a><ul>
<li><a class="reference internal" href="#overview">Overview</a></li>
<li><a class="reference internal" href="#dependencies">Dependencies</a></li>
<li><a class="reference internal" href="#modules-attributes-available">Modules/Attributes available</a></li>
<li><a class="reference internal" href="#examples">Examples</a><ul>
<li><a class="reference internal" href="#using-phyml-with-relaxphylip">Using PhyML with RelaxPhylip</a></li>
<li><a class="reference internal" href="#using-ete3paml">Using ETE3PAML</a></li>
<li><a class="reference internal" href="#using-the-filteredtree-implementation-of-iqtree">Using the FilteredTree implementation of IQTree</a></li>
</ul>
</li>
</ul>
</li>
</ul>

  </div>
<search id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" autocomplete="off" autocorrect="off" autocapitalize="off" spellcheck="false"/>
      <input type="submit" value="Go" />
    </form>
    </div>
</search>
<script>document.getElementById('searchbox').style.display = "block"</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &#169;2017, Shaurita Hutchins & Robert Gilmore.
      
      |
      Powered by <a href="https://www.sphinx-doc.org/">Sphinx 8.2.3</a>
      &amp; <a href="https://alabaster.readthedocs.io">Alabaster 1.0.0</a>
      
      |
      <a href="../_sources/orthologs/phylogeneticsreadme.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    
    <a href="https://github.com/Datasnakes/OrthoEvolution" class="github">
        <img src="../_static/github-banner.svg" alt="Fork me on GitHub" class="github"/>
    </a>
    

    
  </body>
</html>