<!DOCTYPE html>

<html lang="python" data-content_root="../">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>Contents of the /blast/db/ directory &#8212; OrthoEvolution 1.0.0b2 documentation</title>
    <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=fa44fd50" />
    <link rel="stylesheet" type="text/css" href="../_static/basic.css?v=b08954a9" />
    <link rel="stylesheet" type="text/css" href="../_static/alabaster.css?v=27fed22d" />
    <script src="../_static/documentation_options.js?v=3e4b6f43"></script>
    <script src="../_static/doctools.js?v=9bcbadda"></script>
    <script src="../_static/sphinx_highlight.js?v=dc90522c"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
   
  <link rel="stylesheet" href="../_static/custom.css" type="text/css" />
  

  
  

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <section id="contents-of-the-blast-db-directory">
<h1>Contents of the /blast/db/ directory<a class="headerlink" href="#contents-of-the-blast-db-directory" title="Link to this heading">Â¶</a></h1>
<p>The names of these databases and their contents are listed below.</p>
<table class="docutils align-default">
<thead>
<tr class="row-odd"><th class="head"><p>File Name</p></th>
<th class="head"><p>Content Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>16SMicrobial.tar.gz</p></td>
<td><p>Bacterial and Archaeal 16S rRNA sequences
from BioProjects 33175 and 33117</p></td>
</tr>
<tr class="row-odd"><td><p>FASTA/</p></td>
<td><p>Subdirectory for FASTA formatted sequences</p></td>
</tr>
<tr class="row-even"><td><p>README</p></td>
<td><p>README for this subdirectory (this file)</p></td>
</tr>
<tr class="row-odd"><td><p>Representative_Genomes.*t
ar.gz</p></td>
<td><p>Representative bacterial/archaeal genomes
database</p></td>
</tr>
<tr class="row-even"><td><p>cdd_delta.tar.gz</p></td>
<td><p>Conserved Domain Database sequences for use
with stand alone deltablast</p></td>
</tr>
<tr class="row-odd"><td><p>cloud/</p></td>
<td><p>Subdirectory of databases for BLAST AMI; see
<a class="reference external" href="http://1.usa.gov/TJAnEt">http://1.usa.gov/TJAnEt</a></p></td>
</tr>
<tr class="row-even"><td><p>env_nr.*tar.gz</p></td>
<td><p>Protein sequences for metagenomes</p></td>
</tr>
<tr class="row-odd"><td><p>env_nt.*tar.gz</p></td>
<td><p>Nucleotide sequences for metagenomes</p></td>
</tr>
<tr class="row-even"><td><p>est.tar.gz</p></td>
<td><p>This file requires est_human.<em>.tar.gz,
est_mouse.</em>.tar.gz, and
est_others.*.tar.gz files to function. It
contains the est.nal alias so that searches
against est (-db est) will include
est_human, est_mouse and est_others.</p></td>
</tr>
<tr class="row-odd"><td><p>est_human.*.tar.gz</p></td>
<td><p>Human subset of the est database from the
est division of GenBank, EMBL and DDBJ.</p></td>
</tr>
<tr class="row-even"><td><p>est_mouse.*.tar.gz</p></td>
<td><p>Mouse subset of the est databasae</p></td>
</tr>
<tr class="row-odd"><td><p>est_others.*.tar.gz</p></td>
<td><p>Non-human and non-mouse subset of the est
database</p></td>
</tr>
<tr class="row-even"><td><p>gss.*tar.gz</p></td>
<td><p>Sequences from the GSS division of GenBank,
EMBL, and DDBJ</p></td>
</tr>
<tr class="row-odd"><td><p>htgs.*tar.gz</p></td>
<td><p>Sequences from the HTG division of GenBank,
EMBL, and DDBJ</p></td>
</tr>
<tr class="row-even"><td><p>human_genomic.*tar.gz</p></td>
<td><p>Human RefSeq (NC_######) chromosome records
with gap adjusted concatenated NT_ contigs</p></td>
</tr>
<tr class="row-odd"><td><p>nr.*tar.gz</p></td>
<td><p>Non-redundant protein sequences from
GenPept, Swissprot, PIR, PDF, PDB, and NCBI
RefSeq</p></td>
</tr>
<tr class="row-even"><td><p>nt.*tar.gz</p></td>
<td><p>Partially non-redundant nucleotide sequences
from all traditional divisions of GenBank,
EMBL, and DDBJ excluding GSS, STS, PAT, EST,
HTG, and WGS.</p></td>
</tr>
<tr class="row-odd"><td><p>other_genomic.*tar.gz</p></td>
<td><p>RefSeq chromosome records (NC_######) for
non-human organisms</p></td>
</tr>
<tr class="row-even"><td><p>pataa.*tar.gz</p></td>
<td><p>Patent protein sequences</p></td>
</tr>
<tr class="row-odd"><td><p>patnt.*tar.gz</p></td>
<td><p>Patent nucleotide sequences. Both patent
databases are directly from the USPTO, or
from the EPO/JPO via EMBL/DDBJ</p></td>
</tr>
<tr class="row-even"><td><p>pdbaa.*tar.gz</p></td>
<td><p>Sequences for the protein structure from the
Protein Data Bank</p></td>
</tr>
<tr class="row-odd"><td><p>pdbnt.*tar.gz</p></td>
<td><p>Sequences for the nucleotide structure from
the Protein Data Bank. They are NOT the
protein coding sequences for the
corresponding pdbaa entries.</p></td>
</tr>
<tr class="row-even"><td><p>refseq_genomic.*tar.gz</p></td>
<td><p>NCBI genomic reference sequences</p></td>
</tr>
<tr class="row-odd"><td><p>refseq_protein.*tar.gz</p></td>
<td><p>NCBI protein reference sequences</p></td>
</tr>
<tr class="row-even"><td><p>refseq_rna.*tar.gz</p></td>
<td><p>NCBI Transcript reference sequences</p></td>
</tr>
<tr class="row-odd"><td><p>sts.*tar.gz</p></td>
<td><p>Sequences from the STS division of GenBank,
EMBL, and DDBJ</p></td>
</tr>
<tr class="row-even"><td><p>swissprot.tar.gz</p></td>
<td><p>Swiss-Prot sequence database (last major
update)</p></td>
</tr>
<tr class="row-odd"><td><p>taxdb.tar.gz</p></td>
<td><p>Additional taxonomy information for the
databases listed here providing common and
scientific names</p></td>
</tr>
<tr class="row-even"><td><p>tsa_nt.*tar.gz</p></td>
<td><p>Sequences from the TSA division of GenBank,
EMBL, and DDBJ</p></td>
</tr>
<tr class="row-odd"><td><p>vector.tar.gz</p></td>
<td><p>Vector sequences from 2010, see Note 2 in
section 4.</p></td>
</tr>
</tbody>
</table>
</section>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="Main">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="../index.html">OrthoEvolution</a></h1>



<p class="blurb">An easy to use and comprehensive python package which aids in the analysis and visualization of comparative evolutionary genetics</p>




<p>
<iframe src="https://ghbtns.com/github-btn.html?user=Datasnakes&repo=OrthoEvolution&type=watch&count=true&size=large&v=2"
  allowtransparency="true" frameborder="0" scrolling="0" width="200px" height="35px"></iframe>
</p>






<search id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" autocomplete="off" autocorrect="off" autocapitalize="off" spellcheck="false"/>
      <input type="submit" value="Go" />
    </form>
    </div>
</search>
<script>document.getElementById('searchbox').style.display = "block"</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &#169;2017, Shaurita Hutchins & Robert Gilmore.
      
      |
      Powered by <a href="https://www.sphinx-doc.org/">Sphinx 8.2.3</a>
      &amp; <a href="https://alabaster.readthedocs.io">Alabaster 1.0.0</a>
      
      |
      <a href="../_sources/tools/ncbireadme.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    
    <a href="https://github.com/Datasnakes/OrthoEvolution" class="github">
        <img src="../_static/github-banner.svg" alt="Fork me on GitHub" class="github"/>
    </a>
    

    
  </body>
</html>