# Phylogenetics Documentation

This document will provide information and guidelines about using clustal omega and PAML \(with ete3\) in order to generate alignments and dN/dS data about orthologous genes.

## Clustal Omega

### Format Examples

**Directory name:**  `cds`

**CDS input filename:** `HTR1A_Homo_sapiens_cds_nucl.fasta`

**Fasta file format:**

`>H_sapiens`
`ATGGATGTGGTCAACAGCCTTCTTGTGAATGGAAGCAACATCACCCCTCCTTGCGAACTAGGCATCGAAAATGAGACGCTTTTCTGCTTGGATCAGCCCCATTCATCCAAAGAGTGGCAGCCAGCTGTGCAGATTCTCTTGTATTCCTTGATATTCCTGCTCAGCGTGCTGGGGAACACGCTGGTCATCACGGTGCTGATTCGAAACAAGAGGATGAGAACCGTCACCAACATCTTCCTGCTGTCCCTGGCCATCAGCGACCTCATGCTCTGCCTCTTCTGTATGCCGTTCAACCTCATCCCCAACCTGCTCAAGGATTTCATCTTCGGGAACGCTGTTTGCAAGACCACCACCTACTTCATGGGCACCTCTGTGAGTGTATCCACTTTTAATCTGGTCGCCATATCTCTGGAGCGATACGGTGCGATTTGCAAACCCTTACAGTCCCGGGTCTGGCAGACGAAATCCCACGCTTTGAAGGTGATTGCTGCTACCTGGTGCCTCTCCTTTACCATCATGACTCCGTACCCAATTTATAGCAACTTGGTGCCTTTTACCAAAAATAACAACCAGACCGCAAATATGTGCCGCTTTCTACTGCCAAATGATGTTATGCAGCAGTCCTGGCACACGTTCCTGTTACTCATCCTCTTTCTTATTCCTGGAATCGTAATGATGGTAGCATATGGATTAATCTCTTTGGAACTTTACCAAGGAATAAAATTTGATGCTAGCCAGAAGAAGTCTGCTAGAGAAAGGAAACCGAGCGCCGGCAGCAGCGGCCGATACGAGGACAGTGACGGGTGTTACCTGCAGAAGTCCAAGCACCCACGCAAGCTGGAGCTTCAGCGCCTGTCGCCCGGCGGCAGCGGCAGGGTCAACCGCATCAGGAGCAGCAGCTCCGCGGCCAACCTGATGGCCAAGAAGCGGGTGATCCGCATGCTCATGGTCATCGTGGTCCTCTTCTTCCTGTGCTGGATGCCCATCTTCAGCGTCAACGCCTGGCGGGCCTATGACACGGCCTCCGCCGAGCGCCACCTCTCGGGGACCCCCATTTCCTTCATCCTCCTGCTCTCCTACACCTCCTCCTGCGTCAACCCCATCATCTACTGCTTCATGAACAAACGGTTCCGCCTCGGCTTCATGGCCACCTTCCCCTGCTGCCCCAACCCTGGTCCCCCAGGGGCGAGAGGAGAGCCGGGAGAGGAGGAGGAAGGCAGGACCACCGGGGCCTCGCTGTCCAGGTACTCATACAGCCACATGAGCGCCTCTGCCCCGCCCCCGTGA`

**CDS output file name:** `HTR1A_cds_nucl_aligned.fasta`

### Analysis Parameters

`infile="MASTER_" + gene + "_CDS1.ffn",`

`outfile=gene + "_aligned_cds_nucl.fasta",`

`seqtype="DNA",`

`max_hmm_iterations=2,`

`infmt="fasta", `

`outfmt="fasta",`

`iterations=3,`

`verbose=True,`

`threads=8,`

`force=True,`

`log=gene + "_alignment.log"`

### _Notes_

2 spaces should be after the species/common name in the fasta file.

The input and output file will be in fasta format.

## ETE3 & PAML

### Format Examples

**Directory name:** `M1.HTR1A`

**Input fasta file name:** `HTR1A_Homo_sapiens_cds_nucl.fasta`

**Input tree file name:** `species_tree.nw`

**Output filenames: **_Output filenames will be the same as PAML's standard output._

